### ISSUE TITLE
Not able to use NotoColorEmoji font

### ISSUE THREAD

**Issue by yiminli** [ocornut/imgui#6302]
```
Dear ImGui 1.89.5 WIP (18948)
--------------------------------
sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=201703
define: __linux__
define: __GNUC__=9
--------------------------------
io.BackendPlatformName: imgui_impl_sdl2
io.BackendRendererName: imgui_impl_opengl3
io.ConfigFlags: 0x00000003
 NavEnableKeyboard
 NavEnableGamepad
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000000E
 HasMouseCursors
 HasSetMousePos
 RendererHasVtxOffset
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1959.00,921.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

**My Issue/Question:**
I want to use font NotoColorEmoji_WindowsCompatible.ttf and display a smile face similar to what is shown in:
https://github.com/ocornut/imgui/issues/4566#issuecomment-924803382
It seems the font is not loaded correctly.

**Screenshots/Video**
![image](https://user-images.githubusercontent.com/1844443/229997149-bd2e0d47-65f6-4a53-ae8b-d45ba0beb1c3.png)

**Standalone, minimal, complete and verifiable example:** _(see )_
```
diff main.cpp build/imgui/examples/example_sdl2_opengl3/main.cpp
6,8c6
< #include <string>
< #include "misc/freetype/imgui_freetype.h"
<
---
> #include "imgui.h"
105,116d102
<     io.Fonts->AddFontDefault();
<     {
<         static ImWchar ranges[] = { 0x1, 0x1FFFF, 0 };
<         static ImFontConfig cfg;
<         cfg.OversampleH = cfg.OversampleV = 1;
<         cfg.MergeMode = true;
<         cfg.FontBuilderFlags |= ImGuiFreeTypeBuilderFlags_LoadColor;
<         const std::string icon_font = "/tmp/NotoColorEmoji_WindowsCompatible.ttf";
<         io.Fonts->AddFontFromFileTTF(icon_font.c_str(), 109.0f, &cfg, ranges);
<         io.Fonts->Build();
<     }
<
161,166d146
<
<             ImGui::Begin("font test");
<             static char str[128] = "😊Hello, world!";
<             ImGui::Text("%s", str);
<             ImGui::InputText("input text", str, IM_ARRAYSIZE(str));
<             ImGui::End();
```
Full example with CMake (run build_and_run.sh after unzipping):
[emoji_font.zip](https://github.com/ocornut/imgui/files/11155791/emoji_font.zip)

**Comment by ocornut** [comment]
Two things you can do to begin debugging this:
- Browse `Fonts->Input 1 (Noto etc)` in that debug window and see if the glyph is listed and visible. This is to validate that it has been rendered properly.
- Call `DebugTextEncoding(str)` with your string containing an icon. This is to validate that you compiler UTF-8 encoding settings are correct. See what are the decoded value if they match the codepoint of that Emoji. A `"😊Hello, world!"` literal is NOT standard and requires per-compiler settings to be UTF-8 encoded You can use `(const char*)u8"😊Hello, world!"` as a more portable way to UTF-8 encode, which is best wrapped into a macro to make things shorter.

**Comment by yiminli** [comment]
Thanks for reminding me about u8 string literal. I updated the code to:
            static char str[128] = u8"😊Hello, world!";

😊 is U+1F60A. Seems rendering of the glyph has some problem.
![image](https://user-images.githubusercontent.com/1844443/230287872-92541cd9-7c8b-4ad3-8186-3fb9fb97f444.png)

**Comment by ocornut** [comment]
I can confirm this on my side.
I don't know if it is a bug/unsupported feature in FreeType or if we are misusing Freetype.
It works for me with "seguiemj.tff" but not with "NotoColorEmoji.ttf" neither "NotoColorEmoji_WindowsCompatible.ttf".

Some references:
https://stackoverflow.com/questions/54781525/loading-colored-fonts-with-freetype

**Comment by ItsBranK** [comment]
Man I've been spending the last 3 days trying to get this same font to work, finally found this and saw it was an issue...fml

**Comment by yiminli** [comment]
clfontpng (https://gist.github.com/jokertarot/7583938) can render emoji in NotoColorEmoji_WindowsCompatible.ttf into png file without a problem.
The author used FreeType 2.5. I tested it working with FreeType 2.12.1:
./clfontpng NotoColorEmoji_WindowsCompatible.ttf 🥳

**Comment by batcholi** [comment]
Any update on this issue?

**Comment by sakiodre** [comment]
Apparently, the notoemoji font is not scalable (`FT_IS_SCALABLE(face)` is false), making [`FT_Request_Size`](https://freetype.org/freetype2/docs/reference/ft2-sizing_and_scaling.html#ft_request_size) defaults to 0 for width and height, this can be fixed by using [`FT_Select_Size`](https://freetype.org/freetype2/docs/reference/ft2-sizing_and_scaling.html#ft_select_size) instead, but that function only selects a fixed size, making the "font size" meaningless.

![image](https://github.com/ocornut/imgui/assets/136492105/63894dc2-a566-4be5-ba59-1e55e45095a4)

To reproduce:
```cpp
cfg.FontBuilderFlags |= ImGuiFreeTypeBuilderFlags_LoadColor | ImGuiFreeTypeBuilderFlags_Bitmap;
```
```cpp

    bool FreeTypeFont::IsCBDTFont()
    {
        FT_ULong tag = FT_MAKE_TAG('C', 'B', 'D', 'T');
        FT_ULong length = 0;
        FT_Load_Sfnt_Table(Face, tag, 0, nullptr, &length);
        return length > 0;
    }

    void FreeTypeFont::SetPixelHeight(int pixel_height)
    {
        if (IsCBDTFont())
        {
            IM_ASSERT(Face->num_fixed_sizes != 0);
            LoadFlags |= FT_LOAD_COLOR;

            int best_match = 0;
            int diff = std::abs(pixel_height - Face->available_sizes[0].width);
            for (int i = 1; i < Face->num_fixed_sizes; ++i)
            {
                int ndiff = std::abs(pixel_height - Face->available_sizes[i].width);
                if (ndiff < diff)
                {
                    best_match = i;
                    diff = ndiff;
                }
            }

            FT_Select_Size(Face, best_match);
        }
        else
        {
            // Vuhdo: I'm not sure how to deal with font sizes properly. As far as I understand, currently ImGui assumes that the 'pixel_height'
            // is a maximum height of an any given glyph, i.e. it's the sum of font's ascender and descender. Seems strange to me.
            // NB: FT_Set_Pixel_Sizes() doesn't seem to get us the same result.
            FT_Size_RequestRec req;
            req.type = (UserFlags & ImGuiFreeTypeBuilderFlags_Bitmap) ? FT_SIZE_REQUEST_TYPE_NOMINAL : FT_SIZE_REQUEST_TYPE_REAL_DIM;
            req.width = 0;
            req.height = (uint32_t)pixel_height * 64;
            req.horiResolution = 0;
            req.vertResolution = 0;
            FT_Request_Size(Face, &req);
        }

        // Update font info
        FT_Size_Metrics metrics = Face->size->metrics;
        Info.PixelHeight = (uint32_t)pixel_height;
        Info.Ascender = (float)FT_CEIL(metrics.ascender);
        Info.Descender = (float)FT_CEIL(metrics.descender);
        Info.LineSpacing = (float)FT_CEIL(metrics.height);
        Info.LineGap = (float)FT_CEIL(metrics.height - metrics.ascender + metrics.descender);
        Info.MaxAdvanceWidth = (float)FT_CEIL(metrics.max_advance);
    }
```

I can't figure out how to adjust the scale, given that it's "unscalable" but as a workaround, we can use the SVGinOT font [noto-untouchedsvg.ttf](https://github.com/googlefonts/color-fonts/blob/main/fonts/noto-untouchedsvg.ttf) with my PR #6591

![image](https://github.com/ocornut/imgui/assets/136492105/a0365b6e-96b0-46b4-88ed-a8ef5d532093)

**Comment by sakiodre** [comment]
@ocornut I think we can close this as this is freetype's lack of support for CBDT & COLRv1 fonts (see the technical details I posted above).

With `IMGUI_ENABLE_FREETYPE_OTSVG` we can use the [noto-untouchedsvg.ttf](https://github.com/googlefonts/color-fonts/blob/main/fonts/noto-untouchedsvg.ttf) variant instead.

**Comment by ocornut** [comment]
Thank you @sakiodre, closing this indeed!

**Comment by LordOfDragons** [comment]
Any plans on supporting COLRv1? I'm using the latest FreeType and when I generate a COLRv1 for my project it is all empty (including using the font in the system... but interestingly not if using LibreWriter) but when I generate COLRv0 it works. I would like to use COLRv1 if possible even if it is right now not a show-stopper.

**Comment by Green-Sky** [comment]
I don't think this should be closed. Everyone else (eg ff, chrome) can do it too.
I do not see why Dear imgui would be special.

> Apparently, the notoemoji font (CBDT variant) is not scalable (`FT_IS_SCALABLE(face)` is false), making [`FT_Request_Size`](https://freetype.org/freetype2/docs/reference/ft2-sizing_and_scaling.html#ft_request_size) defaults to 0 for width and height, this can be fixed by using [`FT_Select_Size`](https://freetype.org/freetype2/docs/reference/ft2-sizing_and_scaling.html#ft_select_size) instead, but that function only selects a fixed size, making the "font size" meaningless.

I took a quick look at the code, and we currently silently ignore the `FT_Request_Size` returned error too. But something like this needs to be incorporated.

**Comment by ocornut** [comment]
> Everyone else (eg ff, chrome) can do it too. I do not see why Dear imgui would be special.

Because Dear ImGui do not use many third-party libraries and is not a web browser.

The project description is "Dear ImGui is designed to enable fast iterations and to empower programmers to create content creation tools and visualization / debug tools (as opposed to UI for the average end-user). "

I closed based on the post above stating that FreeType does not support CBDT & COLRv1 fonts.
If you can come with an implementation/fix that either doesn't drag more dependency, either makes them optional, feel free to make a pull request.

**Comment by Green-Sky** [comment]
> I closed based on the post above stating that FreeType does not support CBDT & COLRv1 fonts. If you can come with an implementation/fix that either doesn't drag more dependency, either makes them optional, feel free to make a pull request.

FreeType does give you an api to extract all the paint instructions, which is then consumed by eg cairo(ff) or skia(chrome).
ImGui chose to use plutosvg/plutovg, which I also think would be the place for this to be implemented, so I opend https://github.com/sammycage/plutosvg/issues/40 over there.

I still think "closed" is not the correct state for this issue though, but you do you (: .

**Comment by Green-Sky** [comment]
> FreeType does not support CBDT

Now that one is just not true. I think we are just using the api wrong. Like how we have to select the size instead of "request" the size, like I stated before.
Bitmaps have the downside to always be the same size, so we will have to decide if we downscale them to the current font size or keep them full sized in vram (probably not).

(relates to https://github.com/ocornut/imgui/issues/8190 and others)

**Comment by ocornut** [comment]
> I opened https://github.com/sammycage/plutosvg/issues/40 over there.

Thanks. If it can be supported in PlutoSVG then supposedly it wouldn't need an update on dear imgui side?

> I still think "closed" is not the correct state for this issue though, but you do you (: .

For semi-fringe features related to tech that are more foreign to me, I am forced to put them aside to help me prioritize on core Dear ImGui features, otherwise it is humanly unmanageable to me. 

I'm really not sure what to do with this, I'm not equipped for it and it outside my priority zone. I'm however absolutely OK to merge support/improvements/fixes PR if the code is sensible and well written. 

It if end up extracting bitmaps they should be downscaled to current font size probably.

**Comment by Green-Sky** [comment]
> > I opened [sammycage/plutosvg#40](https://github.com/sammycage/plutosvg/issues/40) over there.
> 
> Thanks. If it can be supported in PlutoSVG then supposedly it wouldn't need an update on dear imgui side?

No, there is no simple, preexisting API for this. For svg, freetype has a hook that can be set. So expect a little friction here. But as I state in that issue, there are other downstream projects that would benefit form such an api (I name SDL_ttf there).

> > I still think "closed" is not the correct state for this issue though, but you do you (: .
> 
> For semi-fringe features related to tech that are more foreign to me, I am forced to put them aside to help me prioritize on core Dear ImGui features, otherwise it is humanly unmanageable to me.
> 
> I'm really not sure what to do with this, I'm not equipped for it and it outside my priority zone. I'm however absolutely OK to merge support/improvements/fixes PR if the code is sensible and well written.

yay

> It if end up extracting bitmaps they should be downscaled to current font size probably.

I think so too. The new font system accounts for dpi stuff, right?

Also to add to this, imgui has had support for CBDT since 2021 https://github.com/ocornut/imgui/pull/3879 , it just only works when the fontsize is exactly one of the font's provided bitmap sizes.

Here is me loading the system provided noto color emojies font at a pixelsize of 109:

<img width="251" height="94" alt="Image" src="https://github.com/user-attachments/assets/dadb87d3-ea99-439b-9575-594dc24edb0c" />

edit: Also I have been following https://github.com/ocornut/imgui/pull/8857 , as it tangates this somewhat.

**Comment by Green-Sky** [comment]
There is now a PR that implements bitmap font size selection and downscaling. This should make it possible to use NotoColorEmoji and AppleColorEmoji in most scenarios.

https://github.com/ocornut/imgui/pull/9012

Please test :)

---

If that PR is merged, I consider this issue done. COLRv1 is a lower priority and depends on external dependencies. A new issue or PR for COLRv1 can be opened when that is ready.

**Comment by NsPro04** [comment]
In addition to all the planned font changes, are there any plans to add support for combined characters?

For example, the flag <img width="28" height="15" alt="Image" src="https://github.com/user-attachments/assets/0f25b210-6352-4168-82b6-a662ce47039f" /> (🇺🇸) is a combination of (🇺) and (🇸)
And right now, it's displayed as a combination, not as a flag

This probably isn't quite the right place to ask this question, but anyway

<img width="213" height="140" alt="Image" src="https://github.com/user-attachments/assets/b8aa2c5c-0ab4-461a-82e4-ecfcf2ecac81" />

**Comment by ocornut** [comment]
> In addition to all the planned font changes, are there any plans to add support for combined characters?

No plan to support advanced text shaping and Unicode features, no.

> This probably isn't quite the right place to ask this question, but anyway

It is not the right place to ask this question indeed.

**Comment by ocornut** [comment]
An extension to stb_truetype.h was released today. It seems like we could easily add opt-in support via a define.


https://mastodon.gamedev.place/@MikkoMononen/116880349143031654
https://codeberg.org/memononen/colr_truetype

**Comment by alan-liu-agesx** [comment]
在配置了 IMGUI_ENABLE_FREETYPE 和 IMGUI_USE_WCHAR32 的情况下，以下这个字体可以使用。
它（Twemoji.Mozilla.ttf）和 seguiemj.ttf 一样，属于 COLRv1 版本。
https://github.com/mozilla/twemoji-colr/releases

**Comment by Turtle-PB** [comment]
Check my github I proposed a fix.
________________________________

**Comment by Turtle-PB** [comment]
@yiminli I've implemented a fix for this based on @sakiodre's analysis — PR #9483.

**The fix:** Detect the CBDT table via `FT_Load_Sfnt_Table()`, and when present, use `FT_Select_Size()` to pick the closest bitmap strike instead of `FT_Request_Size()`, plus set `FT_LOAD_COLOR`:

```cpp
// Detect CBDT/COLR bitmap fonts (NotoColorEmoji, Twemoji COLR, etc.)
FT_ULong cbdt_tag = FT_MAKE_TAG('C', 'B', 'D', 'T');
FT_ULong cbdt_length = 0;
FT_Error cbdt_err = FT_Load_Sfnt_Table(face, cbdt_tag, 0, nullptr, &cbdt_length);
bool is_cbdt_font = (cbdt_err == 0 && cbdt_length > 0);

if (is_cbdt_font && face->num_fixed_sizes > 0)
{
    // Select the closest available bitmap strike size
    int best_match = 0;
    FT_Int best_diff = INT_MAX;
    FT_Int target_size = (FT_Int)(size * rasterizer_density);
    for (int i = 0; i < face->num_fixed_sizes; i++)
    {
        FT_Int diff = (FT_Int)face->available_sizes[i].height - target_size;
        if (diff < 0) diff = -diff;
        if (diff < best_diff) { best_diff = diff; best_match = i; }
    }
    FT_Select_Size(face, best_match);
    LoadFlags |= FT_LOAD_COLOR;
}
else
{
    // Standard scalable font path (unchanged)
    FT_Request_Size(face, &req);
}
```

**Key properties:**
- Scalable fonts (TTF/OTF) take the exact same code path as before — no behavioral change
- No new dependencies — existing FreeType API only
- Works with `NotoColorEmoji_WindowsCompatible.ttf` and Mozilla's [Twemoji COLR v0.7.0](https://github.com/mozilla/twemoji-colr/releases)

@sakiodre's diagnosis was spot-on — the font isn't scalable so `FT_Request_Size` defaults to 0x0. One note for anyone using this: since CBDT strikes come in fixed sizes (e.g. 128px), the loaded glyphs are rasterized at the nearest available strike, so very small requested sizes will downscale from the bitmap rather than rasterize cleanly.

@ocornut — the PR keeps everything within the existing FreeType backend, no new deps. Happy to adjust the approach if you'd prefer a different detection method (e.g. checking `FT_IS_SCALABLE` directly instead of the CBDT table probe).

**Comment by Green-Sky** [comment]
@Turtle-PB This reads fully generated by some llm. Also your pr implements a subset of the existing pr #9012

**Comment by ocornut** [comment]
@Turtle-PB For avoidance of doubt please disclose/explain any use of LLM 🙏
